Convert locale files from ini to json format

lunny opened the pull request:
Use JSON format instead of ini format for translations.

silverwind commented:
Very good choice.

wxiaoguang commented:
TBH I won't say it is a good choice. Full screen of `\"`

<img width="2012" height="1294" alt="image" src="https://github.com/user-attachments/assets/128ddd69-ef19-407b-b949-eb1b2d74be22" />

wxiaoguang commented:
By the way, you also need to fix 

* https://github.com/go-gitea/gitea/blob/main/.github/workflows/cron-translations.yml
* `./build/update-locales.sh`

wxiaoguang commented:
> > Full screen of "
> 
> We should eventually remove all HTML from translations which will solve this. There are always ways to interpolate translation into templates without having HTML in the translation itself.

The "eventually has been years, many "eventually" things have been years.

hiifong commented:
I believe switching to TOML would be preferable, as existing translations can be easily converted to the TOML format.

wxiaoguang commented:
> Still I think JSON is a big step forward over INI. It has clear and well-known escaping rules, and any mistake will be found by any JSON parser (which should run on the CI, and this is trivial to add).

Yes, any well-defined format is a big step.

Still think it should clean up the legacy HTML quotes before migrating, to make the code base clean and avoid low-level mistakes.

wxiaoguang commented:
> > Still I think JSON is a big step forward over INI. It has clear and well-known escaping rules, and any mistake will be found by any JSON parser (which should run on the CI, and this is trivial to add).
> 
> One drawback of JSON format is that it does not support adding comments.

Agree, that's also a problem.

lunny commented:
> I believe switching to TOML would be preferable, as existing translations can be easily converted to the TOML format, Nor will you see a screen full of `\"`

TOML is also a possible choice but it seems weblate haven't support it.
Replace the html tags one by one manually. There are many examples in the template files.

silverwind commented:
I think JSON is a good choice. TOML and YAML are both more much complex specs. I think the downside of having to escape `"` is acceptable.

denyskon commented:
This will definitely be problematic with Crowdin. It doesn't simply support changing the file format, this will be as if we uploaded a completely new set of strings IIRC

lunny commented:
> This will definitely be problematic with Crowdin. It doesn't simply support changing the file format, this will be as if we uploaded a completely new set of strings IIRC

Yes, that the plan. Once this is merged, we can uploaded all the languages files and remove ini format files.

silverwind commented on Makefile:
```diff
@@ -918,14 +918,32 @@ lockfile-check:
+
+.PHONY: translation-check
+translation-check:
+	npm install -g jsonlint find-duplicated-property-keys
```
Add them to devDependencies:

```
pnpm install - D jsonlint find-duplicated-property-keys
```

And  then use them as this:

```
$(NODE_VARS) pnpm exec jsonlint
$(NODE_VARS) pnpm exec find-duplicated-property-keys
```

lunny replied:
d7702ead0e

denyskon commented:
The we'll loose all translations which are created but not proofread ...

denyskon commented:
![Screenshot_20251003_165330_Firefox.jpg](https://github.com/user-attachments/assets/041e6400-e0f1-46d4-9fa1-3247e81f35f0)

What do we do with cases like these?

denyskon reviewed (changes requested):
Blocking until we've dealt with the Crowdin question

lunny commented:
> > ==> options/locale/locale_cs-CZ.json
> > ❌ Invalid JSON syntax: options/locale/locale_cs-CZ.json
> 
> Empty file is not valid json, you need to write `{}` or detect empty file.

The script build/update-locales.sh now checks for empty files.
I also removed the redundant update-translation section from the Makefile.
Additionally, I simplified the Makefile so that the translation check is more concise—it no longer needs a separate shell script.

denyskon commented:
We've given long enough notice, we have to accept that some languages are unmaintained.

denyskon commented:
@lunny please update to latest crowdin state before merging

wxiaoguang commented on tools/convert_locale_ini_json.go:
```diff
@@ -0,0 +1,104 @@
+	"code.gitea.io/gitea/modules/setting"
+)
+
+func convertIniToJSON(data []byte) ([]byte, error) {
```
No need to merge such one-time script into main branch.

lunny commented on Makefile:
```diff
@@ -918,15 +918,21 @@ lockfile-check:
+			echo "Duplicate key found in: $$f"; \
+			exit 1; \
+		fi; \
+	done
```
I think the `find-duplicated-property-keys` is still necessary?

lunny replied:
The script has been removed.

wxiaoguang commented:
> @wxiaoguang any further comments?

My question is whether lunny has a real feasible plan and TODO list for this change, whether he knows what should to be done step by step. According to my knowledge, I guess he doesn't, or even he has, he usually doesn't follow (https://github.com/go-gitea/gitea/issues/36132#issuecomment-3668670098).

But I won't block, it's free for people to do what they want.

wxiaoguang commented on .github/workflows/files-changed.yml:
```diff
@@ -53,7 +53,6 @@ jobs:
               - "Makefile"
               - ".golangci.yml"
               - ".editorconfig"
-              - "options/locale/locale_en-US.ini"
```
why?

lunny commented:
> > After this PR merged, all the JSON files should be uploaded to crowdin immediately.
> 
> Do you have a TODO list? For example: should the GiteaBot be updated for its "I noticed you've updated the locales for non-English languages. " task? [GiteaBot/gitea-backporter@`main`/src/comments.ts#L9](https://github.com/GiteaBot/gitea-backporter/blob/main/src/comments.ts#L9)

Updated the issue content.
